[feat] milestone API 작성

SeunghyoKu opened the pull request:
## 관련 이슈 및 위키
마일스톤 CRUD : #12 #13 #14 #15
## 내용
### [feat] milestone api 작성
- 기존 수원님이 작성해주신 label api 스타일을 참고하여 milestone api를 작성하였습니다.
- **논의하고 싶은 내용**:
    - api 응답할 때 내보내는 message들을 따로 파일을 만들어 관리해볼까요?
    - message style도 통일하는 것은 어떨지.. 여러분의 의견을 듣고 싶습니다. 👀

### [feat] milestone CRUD 관련 query문 작성
> 548616
- insertMilestone | selectMilestone | updateMilestone | deleteMilestone
- 필요한 쿼리문 작성하였습니다.

### [feat] milestone model 작성
> 464aad
- 수원님의 깔꼼한 코드 정리가 너무 이뻐서 적용하였습니다.
- `insert` 동작 같은 경우, 필요한 정보를 하나하나 따로 받아와서 `insertInfo` 라는 배열 안에 넣어 사용하였습니다.
- 가독성을 높이기 위해 위와 같이 사용했는데 보시기에 어떤지 궁금합니다. 😅 (하고 나니깐 애매하더라구요....)

### [feat] milstone API 작성
> 8ac93
- 상태코드를 조금 더 자세히 적고 싶은데 쉽지 않은 것 같습니다.
- 500 에러가 나는 경우가 있을까요..? 😳
- 어제 논의한대로 id 변수는 모두 소문자인 `milestoneid`로 작성하였습니다.

## Why?
- 상태코드를 적용해보고자 했으나... 500 에러만 추가해보았습니다.
- 하지만 500가 서버 오류라고 하나, 어떻게 발생시켜야 좋을지 모르겠어서 500이 일어나는 상황을 테스트해보지는 못했습니다.

SeunghyoKu committed:
[feat] milestone CRUD 관련 query문 작성
- insert, select, update, delete 하는 쿼리문 작성

SeunghyoKu committed:
[feat] milestone router 추가
- 기존에 오류 발생으로 주석처리 되어있던 milestone router를 주석 해제 하였습니다.

SeunghyoKu committed:
[feat] milestone model 작성
- config/query.js에서 sql문을 가져와 insert/select/update/delete 작업

SeunghyoKu committed:
[feat] milstone API 작성
- create: 마일스톤 추가하기
- read: 마일스톤 가져오기
- update: 마일스톤 업데이트하기
- delete: 마일스톤 삭제하기

safience-mia commented on web-BE/models/milestone.js:
```diff
@@ -0,0 +1,38 @@
+const milestoneModel = {
+  insert: async (title, dueDate, content) => {
+    try {
+      const insertInfo = [title, dueDate, content];
```
@j03y14 재윤님이 말씀하셨던 배열 비구조화가 이렇게 작성하는것 같은데 맞나요? 
맞다면 저도 승효님처럼 수정해야겠네요 😺 
저도 이렇게 전달하는게 옳다고 생각해요! 어떤 데이터가 전달되는지 알 수 있으니까요 👍🏻 
_Originally posted by @j03y14 in https://github.com/boostcamp-2020/IssueTracker-12/pull/51#discussion_r514029244_

j03y14 replied:
네 만약에 매개변수로 객체를 받는다면
```javascript
insert : async ({ title, dueDate, content }) => {
  const insertInfo = [title, dueDate, content];
  connection.query(sql.insertMilestone, insertInfo);
}
```
이런식이면 어떨까 했습니다.
승효님께서 하신거는 매개변수로 그냥 변수들을 받아서 승효님께서 하신대로 해도 좋은 것 같아요!👍

SeunghyoKu replied:
재윤님이 말씀하신 방향이 조금 더 사용하긴 편할 것 같긴 해요. 😩
데이터를 넘기기 전에 가공해서 보내주는 게 조금 더 보기 좋을 것 같군요...!

safience-mia commented on web-BE/routes/api/milestone/controller.js:
```diff
@@ -0,0 +1,42 @@
+      milestoneModel.insert(title, dueDate, content);
+      res.status(200).json({ message: '마일스톤 생성 완료' });
+    } catch (err) {
+      res.status(500).json(err);
```
상태코드 논의하면 좋을 것 같아요, 제 단점은 상태코드 관리를 잘 못한다는 점 😅 
이참에 제대로 실습하면 좋을 듯 합니다.

safience-mia commented on web-BE/routes/api/milestone/controller.js:
```diff
@@ -0,0 +1,42 @@
+      milestoneModel.insert(title, dueDate, content);
+      res.status(200).json({ message: '마일스톤 생성 완료' });
+    } catch (err) {
+      res.status(500).json(err);
```
저는 지난 프로젝트 때 모든 api에(read처럼 응답데이터가 따로 있는 경우도) 메세지를 보내긴 했어요! 아래처럼요.

```js
res.status(200).json({ 
  message: '마일스톤 읽기 성공',
  milestoneArr : responseArr 
 });
```

SeunghyoKu replied:
아무 응답 없는 것 보다는 좋을 것 같군요. 그 부분이 고민 됐는데 감사드려요.. !

slogup-mia reviewed (commented):
좋습니다 👍🏻
추후에 마일스톤 read에서 open/close 이슈 개수 가지고 complete percent 계산하는 로직만 잘 짜면 마일스톤 API도 문제 없겠네요!

JungSWon reviewed (approved):
좋습니다 👍🏻
추후에 마일스톤 read에서 open/close 이슈 개수 가지고 complete percent 계산하는 로직만 잘 짜면 마일스톤 API도 문제 없겠네요!

j03y14 commented on web-BE/config/query.js:
```diff
@@ -2,6 +2,14 @@ module.exports = {
+  // milestone
+  insertMilestone:
+    'INSERT INTO Milestone (title,due_date,content) VALUES (?,?,?);',
+  selectMilestone: 'SELECT * FROM Milestone;',
```
별게 아니기는 한데 
```sql
SELECT * FROM Label
```
대신에 어떤 column 값을 받는지 명시해주는 것도 좋은것 같아요!

safience-mia replied:
> 별게 아니기는 한데
> 
> ```sql
> SELECT * FROM Label
> ```
> 
> 대신에 어떤 column 값을 받는지 명시해주는 것도 좋은것 같아요!

ㅋㅋ맞아요 이거 저도 * 이렇게 셀렉했는데 🙄
이거 전에 마스터님인가? 리뷰어님인가 재윤님이랑 똑같이 말씀해주셨어요.

SeunghyoKu replied:
헉 ㅋㅋㅋㅋㅋ 생각하지 못했던 부분인데 보여주는 게 맞는 것 같아요. 😳

j03y14 commented on web-BE/routes/api/milestone/controller.js:
```diff
@@ -0,0 +1,42 @@
+    } = req.body;
+    try {
+      milestoneModel.update(milestoneid, title, dueDate, content);
+      res.status(500).json({ message: '마일스톤 업데이트 완료' });
```
업데이트를 잘 했는데 응답 코드가 500인건 좀 이상한것 같아요🤔

SeunghyoKu replied:
앗... !!!!! 맞네요 ㅋㅋㅋㅋㅋ 소름 돋았어요... 얼른 고치러 갑니다..... 치명적이군요...

j03y14 reviewed (changes requested):
전체적으로 깔끔하고 좋았던 것 같아요!

SeunghyoKu committed:
[refactor] milestone api controller 데이터 받아오는 부분 수정
- milstone api에서 데이터 받아오는 부분을 controller 부분에서 수정하도록 하였습니다.
- controller에서 보다 배열에 담아 저장한 것을
- model에서는 사용만 하는 것이 좋다고 생각되어 고치게 되었습니다.

SeunghyoKu committed:
[style] 쿼리문 가독성 높이기
- 기존에 SELECT * 로 작성하였던 부분을
- 칼럼을 명확히 작성하여 가독성을 높이게 되었습니다.

SeunghyoKu committed:
[style] ESLint 적용
- 헉.. 또 내면의 prettier가 작동해서쌍따옴표로 바뀐 것을 발견하고
- 작은 따옴표로 적용해서올립니다. ㅠㅠ
- ESLint 적용을 자꾸 까먹고 push 하게 되는 군요..
